### PR TITLE
16781 - Migrate Base Address from sbc-common-components

### PR DESCRIPTION
**Pull request by TMunro7**
*Issue #:* /bcgov/entity#16781

*Description of changes:*
Added the BaseAddress and SbcFeeSummary Components from SbcCommonComponents to BcrsCommonComponenets

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

**Comment by seeker25** [comment]
Please change the title, so it includes the ticket number:

EG.
16781 - Migrate Base Address from sbc-common-components

File: src/components/base-address/package.json
Comment (seeker25) [inline src/components/base-address/package.json:12]: Alotta spacing

File: package.json
Comment (seeker25) [inline package.json:23]: What does this do?

**Comment by seeker25** [comment]
We'll need to tweak the existing components so they don't point to sbc-common-components and point to this component instead.

File: src/mixins/countries-provinces-mixin.ts
Comment (seeker25) [inline src/mixins/countries-provinces-mixin.ts:5]: Just a note, this should be fixed before merging

**Review by seeker25 — APPROVED** [review]
Waiting for CI to pass + maybe a check over from Severin (plus release)

File: src/mixins/countries-provinces-mixin.ts
Comment (JazzarKarim) [inline src/mixins/countries-provinces-mixin.ts:7]: This actually is a game changer. That makes my life so much easier.
uncaught error requre

I cannot wait for this PR to be merged 😁

Comment (seeker25) [reply src/mixins/countries-provinces-mixin.ts:7]: lol didn't take long

**Comment by seeker25** [comment]
Minor lint fix to go

File: src/components/base-address/package.json
Comment (severinbeauvais) [inline src/components/base-address/package.json:13]: Do you depend on Vue?

Also please move appropriate packages to "devDependencies".

File: src/components/fee-summary/FeeSummary.vue
Comment (severinbeauvais) [inline src/components/fee-summary/FeeSummary.vue:67]: These really _should_ import the shared items.

This might work here (in SB) but probably won't work when imported into the host app.

File: src/components/fee-summary/FeeSummary.vue
Comment (severinbeauvais) [inline src/components/fee-summary/FeeSummary.vue:68]: These really _should_ import the shared items.

This might work here (in SB) but probably won't work when imported into the host app.

File: src/components/fee-summary/fee.ts
Comment (severinbeauvais) [inline src/components/fee-summary/fee.ts:8]: Why not move this (and the other) interfaces to the interfaces package?

File: src/components/fee-summary/fee.services.ts
Comment (severinbeauvais) [inline src/components/fee-summary/fee.services.ts:1]: Should we create a services/ folder and publish that?

File: src/components/fee-summary/constants.ts
Comment (severinbeauvais) [inline src/components/fee-summary/constants.ts:1]: How about moving these to the enums/ package?

Comment (severinbeauvais) [reply src/components/fee-summary/constants.ts:1]: And maybe renaming it to indicate what it's used for (ie, sbc-common-components stuff).

File: tsconfig.json
Comment (severinbeauvais) [inline tsconfig.json:14]: What does this fix? 😃

File: src/components/fee-summary/config-helper.ts
Comment (severinbeauvais) [inline src/components/fee-summary/config-helper.ts:1]: How about moving this to utils/ or a new folder? I'm thinking about reuse (and not in the context of Fee Summary).

File: src/components/fee-summary/common-util.ts
Comment (severinbeauvais) [inline src/components/fee-summary/common-util.ts:1]: Move to utils package?

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [inline src/components/base-address/BaseAddress.vue:18]: I think that if you run `npm run lint`, it will auto-fix this. But watch for undesired changes.

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [inline src/components/base-address/BaseAddress.vue:153]: Should be package folders (`@...`) not local paths.

File: src/components/base-address/BaseAddress.stories.ts
Comment (severinbeauvais) [inline src/components/base-address/BaseAddress.stories.ts:1]: See other stories for updated code (minor). Meta, etc.

File: tsconfig.json
Comment (seeker25) [reply tsconfig.json:14]: Importing JSON via `import`

File: src/components/fee-summary/fee.ts
Comment (TMunro7) [reply src/components/fee-summary/fee.ts:8]: while moving the interfaces over to the interfaces package I noticed that the FilingData.ts and the existing filing-data-interface.ts are very similar. which one should I use?

Comment (severinbeauvais) [reply src/components/fee-summary/fee.ts]: They don't look like the same object to me.

However, your 2 interfaces look similar. Are they the same object (used for the same purpose)?
- Fee
- PayData

**Comment by severinbeauvais** [comment]
I recommend a separate PR for the Storybook files, so that I don't have to see 106 extra files in the diff :)

File: src/interfaces/fee-interface.ts
Comment (severinbeauvais) [inline src/interfaces/fee-interface.ts:1]: Please name this FeeIF.

File: src/interfaces/pay-data-interface.ts
Comment (severinbeauvais) [inline src/interfaces/pay-data-interface.ts:2]: Please name this PayDataIF.

**Comment by seeker25** [comment]
> mmend a separate PR for the Storybook files, so that I don't have to see ~106~ **113** extra files in the diff :)

Yup, came back after one of Tay's commits. I force pushed yesterday to get rid of these

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [inline src/components/base-address/BaseAddress.vue:154]: See if you can import the mixins from the shared package.

Otherwise, the host app will have to provide these mixins in its `@/mixins` folder (possibly via re-import in index file).

File: src/components/fee-summary/SbcFeeSummary.vue
Comment (severinbeauvais) [inline src/components/fee-summary/SbcFeeSummary.vue:62]: Can you import these from the shared packages?

Also update package.json for this component (need to include services).

File: src/services/fee.services.ts
Comment (severinbeauvais) [inline src/services/fee.services.ts:1]: Should we publish this folder?

File: src/utils/index.ts
Comment (severinbeauvais) [inline src/utils/index.ts:2]: Also export from config-helper.

File: src/mixins/validation-mixin.ts
Comment (JazzarKarim) [inline src/mixins/validation-mixin.ts:77]: @seeker25 Travis I'm working on migrating Create UI to Vite. I'm getting an error in `BaseAddress.vue` since the component uses this mixin:
validation mixin error

Any way we can fix this as part of this PR?

Comment (severinbeauvais) [reply src/mixins/validation-mixin.ts:77]: From the error, I'm guessing `maxLength` is null. Does it need to be declared somewhere -- in your code or in the unit test?

Comment (JazzarKarim) [reply src/mixins/validation-mixin.ts:77]: Wait, I think you're right. I'll have a look.

File: src/interfaces/FilingData.ts
Comment (severinbeauvais) [inline src/interfaces/FilingData.ts:1]: Please rename to FilingDataIF.

Also rename file to match.

File: src/interfaces/PayDataIF.ts
Comment (severinbeauvais) [inline src/interfaces/PayDataIF.ts:2]: Please rename to PayDataIF.

File: src/services/index.ts
Comment (severinbeauvais) [inline src/services/index.ts:1]: I (or another repo admin) may need to help you with publishing this folder.

1. add a package.json to this folder
2. add folder to lerna.json
3. add this folder to vite.config.js
4. `lerna bootstrap --hoist`
5. check if this folder got a package-lock.json file
6. commit everything
7. an admin will publish it after your PR is merged

Ref: https://github.com/bcgov/bcrs-shared-components#how-to-create-a-new-package-you-depend-on-eg-new-mixins-folder

Comment (TMunro7) [reply src/services/index.ts]: I followed the first 3 steps and on step 4 I got the message, "'lerna' is not recognized as an internal or external command,
operable program or batch file." when trying to run the command `lerna bootstrap --hoist`

Comment (seeker25) [reply src/services/index.ts]: `npm install -global lerna`

File: src/components/base-address/BaseAddress.vue
Comment (TMunro7) [reply src/components/base-address/BaseAddress.vue:153]: I tried setting the import to `import { ValidationMixin, CountriesProvincesMixin } from '@bcrs-shared-components/mixins/'` and  whenever I launched the storybook I would get the warning: `Neither '@bcrs-shared-components/mixins/.vue' nor '@bcrs-shared-components/mixins/.js(x)' or '@bcrs-shared-components/mixins//index.js(x)' or '@bcrs-shared-components/mixins//index.ts(x)' could be found in 'C:/Users/TAMUNRO/Dev/Forked/bcrs-shared-components/src/components/base-address'`

Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue:153]: Does Storybook import that package (in root package.json)?

Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue:153]: It doesn't. Try it. Or maybe I need to update a config file so SB knows where to find the mixins...

File: src/mixins/validation-mixin.ts
Comment (JazzarKarim) [reply src/mixins/validation-mixin.ts:77]: Unless something comes up, this should be OK. Resolving this conversation.

File: src/services/index.ts
Comment (TMunro7) [reply src/services/index.ts]: image
image
I guess its already done by npm?

File: src/services/package.json
Comment (severinbeauvais) [inline src/services/package.json:3]: Set this to 0.0.0. [ref](https://github.com/bcgov/bcrs-shared-components#how-to-create-a-new-package-you-depend-on-eg-new-mixins-folder)

Also please add the dependencies and devDependencies here (Axios, shared interfaces, etc).

File: src/services/index.ts
Comment (severinbeauvais) [reply src/services/index.ts]: Uninstall the latest Lerna and install this one instead: https://github.com/bcgov/bcrs-shared-components#4-install-dependencies

File: src/components/base-address/BaseAddress.vue
Comment (TMunro7) [reply src/components/base-address/BaseAddress.vue:153]: Added `    "@bcrs-shared-components/mixins": "^1.1.16",` into root package.json and it seems to have worked. no more warning messages

Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue]: Tay, you resolved this conversation that I started but I didn't have a chance to review it.

What is the status of this issue ("should not be local import of mixins")?

Comment (TMunro7) [reply src/components/base-address/BaseAddress.vue]: Thank you for letting me know... The current import uses `import { ValidationMixin, CountriesProvincesMixin } from '@bcrs-shared-components/mixins'` with no visible issues or warnings

File: src/components/fee-summary/package.json
Comment (severinbeauvais) [inline src/components/fee-summary/package.json]: Set this to 0.0.0 also. "Lerna hoist" will take care of updating this (I think).

File: package.json
Comment (severinbeauvais) [inline package.json:14]: I'm glad this works.

Does Storybook need to depend on all te other shared items here?

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue:18]: The GHA warning is now gone, but please run `npm run lint` before committing. It will reformat the HTML but this will ensure consistent style between components/repos.

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue]: Great!

File: src/components/base-address/package.json
Comment (severinbeauvais) [inline src/components/base-address/package.json:3]: This should be 0.0.0. Lerna will update it.

I'm guessing you think v2 is appropriate as this component is newly hosted (this shared repo instead of common components project). I think that's resonable; when Lerna prompts me then I can select v2.0.0. Other reviewers -- thoughts on this?

**Comment by severinbeauvais** [comment]
Please rebase and resolve the merge conflict.

File: src/components/base-address/package.json
Comment (JazzarKarim) [reply src/components/base-address/package.json:3]: I guess making it 0.0.0 is the way to go since that's how we usually create new components in this repo? Lerna would do it's wizardry and updates it.

Comment (TMunro7) [reply src/components/base-address/package.json:3]: Yes, I had it as v2 since it was brought over from common components, but I am fine with changing this to 0.0.0

File: package.json
Comment (TMunro7) [reply package.json:14]: Not sure, but I don't think so.
All other shared items seem to work completely fine without being added here. Not sure why this one was different.

File: src/components/base-address/BaseAddress.vue
Comment (TMunro7) [reply src/components/base-address/BaseAddress.vue:18]: `npm run lint` did not make any changes to the HTML; however, I was able to go through and manually make the styling changes suggested by eslint.

File: src/components/base-address/package.json
Comment (TMunro7) [reply src/components/base-address/package.json:15]: If I had to guess, I would think it depends on Vue, but not entirely sure. Seems to work fine without adding vue to the dependencies, but I can add it to be safe?

File: package.json
Comment (severinbeauvais) [reply package.json:14]: I don't know either so just leave it, then.

Resolving.

File: src/components/base-address/package.json
Comment (severinbeauvais) [reply src/components/base-address/package.json:15]: Yes please. Can't hurt (I think).

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue]: OK, thanks.

Maybe you're not using one of the eslint presets that auto-fixes the HTML...?

I'm not going to worry about this. Resolving.

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [inline src/components/base-address/BaseAddress.vue:173]: This is no longer a local mixin -- please remove comment above.

File: src/components/completing-party/CompletingParty.vue
Comment (severinbeauvais) [inline src/components/completing-party/CompletingParty.vue:150]: Change this to use the shared BaseAddress component. If not then when THIS component is used in the host app then it will look for BaseAddress using the relative path above, and fail.

Also add base-address as a dependency (in package file).

**Review by severinbeauvais — APPROVED** [review]
Thanks for doing this. I wished for this for so long! Haha

Good job 👍

File: src/mixins/validation-mixin.ts
Comment (JazzarKarim) [reply src/mixins/validation-mixin.ts:77]: Can we add optional chaining to maxLength here and to minLength on line 73?

`const min = this.$v[prop][key].$params.minLength?.min`
`const max = this.$v[prop][key].$params.maxLength?.max`

I've tried doing the above on my side and it resolved a bunch of errors.
cc: @seeker25 @severinbeauvais

Comment (severinbeauvais) [reply src/mixins/validation-mixin.ts:77]: But why are those ever null/undefined? Don't we have fallback values for these or are they optional? And it they're optional, why is the code trying to get the min and max value?

(I don't want to hide the issue with conditional chaining if possible.)

Comment (JazzarKarim) [reply src/mixins/validation-mixin.ts:77]: That's the thing. This hasn't been an issue before going to Vite. I'm looking at the code and I don't see the values ever being null or undefined.

Comment (severinbeauvais) [reply src/mixins/validation-mixin.ts:77]:
```
  protected maxLengthRule (prop: string, key: string): boolean | string {
    const max = this.$v[prop][key].$params.maxLength.max
    return Boolean(this.$v[prop] && this.$v[prop][key].maxLength) || `Maximum length is ${max}`
  }
```

So, `$v` is the Vue instance, and Vuelidate has added (or in this case, failed to add?) the right config to the prop/key object. I think those are injected by the following method (which comes from 'vue-plugin-helper-decorator'). Does that decorator work in the latest versions of Vue/Vuetify/vite?

```
export default class BaseAddress extends Mixins(ValidationMixin, CountriesProvincesMixin) {
  /**
   * The validation object used by Vuelidate to compute address model validity.
   * @returns the Vuelidate validations object
   */
  @Validation()
  public validations (): any {
    return { addressLocal: { ...this.schemaLocal } }
  }
```

Are we using the same version of Vuelidate, etc?

Comment (severinbeauvais) [reply src/mixins/validation-mixin.ts:77]: > This hasn't been an issue before going to Vite.

Is changing to Vite the only change or are you also using the shared BaseAddress component instead of the common one?

Comment (JazzarKarim) [reply src/mixins/validation-mixin.ts:77]: The one that's being used is from the sbc-common-components.

Regarding Vuelidate, I'm guessing that's the issue. The version of Vuelidate in Create UI is different from the ones the component is using. I'm going to give this a shot.

**Comment by severinbeauvais** [comment]
Just a couple of small fixes and then lmk when to merge and publish this.

File: src/mixins/validation-mixin.ts
Comment (JazzarKarim) [reply src/mixins/validation-mixin.ts:77]: Yes Sev, agreed. Resolving this conversation.

**Review by JazzarKarim — APPROVED** [review]
LGTM, great stuff Tay 👍

File: vite.config.ts
Comment (severinbeauvais) [inline vite.config.ts:23]: Why do you need this alias for base-address? Isn't it already handled by line 24?

Comment (TMunro7) [reply vite.config.ts:23]: you are right. I realized this addition was redundant after I had pushed

Comment (severinbeauvais) [reply vite.config.ts:23]: Please remove it in next commit.

**Review by severinbeauvais — APPROVED** [review]
Looks good to me.

Merge when ready, then let me know so I can go in and publish this.

**Pull request merged**
